### PR TITLE
Deprecate extension methods for String

### PR DESCRIPTION
**Pull request by franciscodr**

### CODE REVIEW

File: arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt [hunk 1]
```diff
@@ -67,35 +126,109 @@ fun stringFilterIndex(): FilterIndex<String, Int, Char> = FilterIndex { p ->
...
+  "Typeclass instance have been moved to the companion object of the typeclass.",
+  ReplaceWith(
+    "Snoc.string()",
+    "arrow.optics.string", "arrow.optics.typeclasses.Snoc"
+  ),
+  DeprecationLevel.WARNING
+)
+fun String.Companion.snoc(): Snoc<String, Char> = StringSnoc()
 
-fun stringSnoc(): Snoc<String, Char> = Snoc {
```
Comment (nomisRev) [inline arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt:96]: This didn't exist in 0.11.0 👍

File: arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt [hunk 2]
```diff
@@ -67,35 +126,109 @@ fun stringFilterIndex(): FilterIndex<String, Int, Char> = FilterIndex { p ->
...
+  "Typeclass instance have been moved to the companion object of the typeclass.",
+  ReplaceWith(
+    "Cons.string()",
+    "arrow.optics.string", "arrow.optics.typeclasses.Cons"
+  ),
+  DeprecationLevel.WARNING
+)
+fun String.Companion.cons(): Cons<String, Char> = StringCons()
 
-fun stringCons(): Cons<String, Char> = Cons {
```
Comment (nomisRev) [inline arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt:84]: Didn't exist in 0.11.0 👍

File: arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt [hunk 3]
```diff
@@ -67,35 +126,109 @@ fun stringFilterIndex(): FilterIndex<String, Int, Char> = FilterIndex { p ->
...
+  ),
+  DeprecationLevel.WARNING
+)
+fun String.Companion.index(): Index<String, Int, Char> = StringIndex()
 
 /**
  * [Index] instance for [String].
  * It allows access to every [Char] in a [String] by its index's position.
  */
-fun stringIndex(): Index<String, Int, Char> = Index { i ->
```
Comment (nomisRev) [inline arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt:76]: Didn't exist in 0.11.0 👍

File: arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt [hunk 4]
```diff
@@ -36,29 +45,79 @@ fun String.Companion.traversal(): Traversal<String, Char> = object : Traversal<S
...
+  ),
+  DeprecationLevel.WARNING
+)
+fun String.Companion.filterIndex(): FilterIndex<String, Int, Char> = StringFilterIndex()
 
 /**
  * [FilterIndex] instance for [String].
  * It allows filtering of every [Char] in a [String] by its index's position.
  */
-fun stringFilterIndex(): FilterIndex<String, Int, Char> = FilterIndex { p ->
```
Comment (nomisRev) [inline arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt:59]: Didn't exist in 0.11.0 👍

File: arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt [hunk 5]
```diff
@@ -36,29 +45,79 @@ fun String.Companion.traversal(): Traversal<String, Char> = object : Traversal<S
...
+    "Traversal.string()",
+    "arrow.optics.Traversal", "arrow.optics.string"),
+  DeprecationLevel.WARNING
+)
+fun String.Companion.each(): Each<String, Char> = StringEach()
 
 /**
  * [Each] instance for [String].
  */
-fun stringEach(): Each<String, Char> = Each { String.traversal() }
```
Comment (nomisRev) [inline arrow-optics/src/main/kotlin/arrow/optics/extensions/string.kt:44]: Didn't exist in 0.11.0 👍

**Review by nomisRev — APPROVED** [review]
Looks great!! Thanks @franciscodr

**Review by raulraja — APPROVED** [review]
thanks @franciscodr !

**Pull request merged**